server/etcdserver/server.go

```diff
@@ -2155,6 +2147,19 @@ func (s *EtcdServer) monitorVersions() {
 	}
 }
 
+// monitorStorageVersion every monitorVersionInterval updates storage version if needed.
+func (s *EtcdServer) monitorStorageVersion() {
+	monitor := serverversion.NewMonitor(s.Logger(), newServerVersionAdapter(s))
+	for {
+		select {
+		case <-time.After(monitorVersionInterval):
+		case <-s.stopping:
+			return
+		}
+		monitor.UpdateStorageVersionIfNeeded()
```

Comment:
We usually introduce flags to prevent enabling non-backward compatible changes. For example adding a new wal entry that cannot be read by older version. This is not this case as all changes here are backward compatible. Storage Value field will just be ignored by older version. 

As for migration, it is called to update storage version to be up to date with current binary version. It only implements upgrades, and has same safeguards as cluster version for user attempting downgrade. Current implementation can only upgrade the storage after cluster version was changed  (meaning that all cluster members were already upgraded). This is the safe way to maintain storage version as users will do backup before before upgrading any members.

If we would consider introducing feature flags I would propose to put flag to hide downgrade feature. Current downgrade logic doesn't work at all and can lead to breaking cluster. This way storage version will be updated during upgrades, but downgrading will require explicit flag passed.

Reply:
This make sense. Thanks!